utility_code/lib_py3/plugin_data.py

```diff
@@ -0,0 +1,295 @@
+
+    def get_debug_str(self):
+        return str(self)
+
+    def iter_graves(self):
+        graves = self._data.get("graves", [])
+        num_graves = len(graves)
+        for i in range(num_graves):
+            grave = MonumentaGrave(graves[i], self)
+            yield grave
+            graves[i] = grave.serialize()
+
+    def iter_thrown_items(self):
+        thrown_items = self._data.get("thrown_items", [])
+        num_thrown_items = len(thrown_items)
+        for i in range(num_thrown_items):
+            thrown_item = JsonWrappedItem(thrown_items[i], self)
+            yield thrown_item
+            thrown_items[i] = thrown_item.serialize()
+
+    def iter_all_types(self):
+        yield from self.iter_graves()
+        yield from self.iter_thrown_items()
+
+    def iter_block_entities(self):
+        pass
+
+    def iter_entities(self):
+        pass
+
+    def iter_items(self):
+        pass
+
+    def recursive_iter_all_types(self):
+        yield self
+        for obj in self.iter_all_types():
+            yield from obj.recursive_iter_all_types()
+
+    def recursive_iter_block_entities(self):
+        for obj in self.iter_all_types():
+            yield from obj.recursive_iter_block_entities()
+
+    def recursive_iter_entities(self):
+        for obj in self.iter_all_types():
+            yield from obj.recursive_iter_entities()
+
+    def recursive_iter_items(self):
+        for obj in self.iter_all_types():
+            yield from obj.recursive_iter_items()
+
+    def __repr__(self):
+        return f'Graves'
+
+class MonumentaGrave(NbtPathDebug):
+    """A collection of items preserved from a single death."""
+
+    def __init__(self, json_data, parent):
+        self._data = json_data
+
+        self.nbt_path_init(None, parent, parent.root if parent is not None and parent.root is not None else self, None)
+
+        # Cosmetic
+        self._equipment = None
+        equipment = self._data.get("equipment", None)
+        if equipment is not None:
+            self._equipment = {}
+            for key, item_snbt in equipment.items():
+                if item_snbt is None:
+                    self._equipment[key] = None
+                else:
+                    item_nbt = nbt.TagCompound.from_mojangson(item_snbt)
+                    self._equipment[key] = Item(item_nbt)
+
+        # Given to players
+        self._items = None
+        items = self._data.get("items", None)
+        if items is not None:
+            self._items = []
+            for item_json in items:
+                self._items.append(JsonWrappedItem(item_json, self))
+
+    def get_debug_str(self):
+        return str(self)
+
+    def serialize(self):
+        if isinstance(self._equipment, dict):
+            for key, item in self._equipment.items():
+                if item is None:
+                    self._data["equipment"][key] = None
```

Comment:
Is this valid for graves in the plugin code?

Reply:
No, I went and checked too. This was code you wrote :P

I think this is still better than not handling this case though. The plugin looks like it'll never set keys without the matching item, so this block shouldn't get run... I think though that I will take it out, I guess I'd rather have it fail if this did change.

Reply:
Probably best to pop the key, then.